Introduce check to prevent PRs from being merged into non-develop branches

BenHenning opened the pull request:
Introduce a check that prevents PRs from being merged unless they are based on develop. This makes chaining more user-friendly since it prevents accidental merges that would break history. It also provides confidence to code reviewers of chained PRs that they can approve them without worrying about the author merging their code into the base branch (technically the branch could still be changed to develop & merged, but sometimes codeowners will prevent that). This is a reasonable step toward a  reliable chaining mechanism.

Demonstration of the check failing with an arbitrary base branch ([workflow](https://github.com/oppia/oppia-android/pull/2832/checks?check_run_id=2329185121)):
![image](https://user-images.githubusercontent.com/12983742/114483729-bbfd5c80-9bbd-11eb-8195-4fdf4ab8715e.png)

Demonstration of the check passing with a release base branch ([workflow](https://github.com/oppia/oppia-android/pull/2832/checks?check_run_id=2329277518)):
![image](https://user-images.githubusercontent.com/12983742/114483784-ddf6df00-9bbd-11eb-9caf-f4d43a94cce1.png)

To see demonstration of it working with 'develop' see the passing CI for this PR.

I will mark the new workflow as required to enforce the check once the PR is merged. I'll also send out a mail to oppia-android-dev explaining this will require merging in the latest develop changes in order to pass CI.

BenHenning committed:
Update static_checks.yml

Add check to ensure PRs cannot be merged unless they are based on develop.

BenHenning committed:
Update static_checks.yml

Fix syntax error.

seanlip commented:
Hi @BenHenning -- I'm a little confused about what "Request for review" means, since on Oppia-web we only use the Assignees field and ignore the Reviewers field completely. I seem to be a reviewer but not assigned. Do you want me to take a look at this?

(If so, then if it's consistent with Oppia Android norms, please assign me explicitly in future. Thanks!)

BenHenning commented:
@seanlip  'request for review' should behave the same on Oppai Android & web repositories, I believe. It's just the thing that GitHub does automatically when there are codeowners. We use assignees to indicate taking action, so feel free to ignore this until you're explicitly assigned (FWIW you won't be--you were added just because the base branch is causing diffs in CODEOWNERS that are unrelated to the actual change here).

BenHenning committed:
Update static_checks.yml

Test to see if https://github.com/a-b-r-o-w-n/check-base-branch-action/pull/11 addresses https://github.com/a-b-r-o-w-n/check-base-branch-action/issues/10.

BenHenning commented:
Ah weird. @seanlip see comment above. For some reason I at-mentioned myself instead of you in reply.

BenHenning committed:
Update static_checks.yml

Use newer version to see if https://github.com/a-b-r-o-w-n/check-base-branch-action/issues/10 is correct.

BenHenning committed:
Update static_checks.yml

Manually check for base branch rather than using a GitHub action.

BenHenning committed:
Update static_checks.yml

Remove old actions check.

BenHenning committed:
Update static_checks.yml

Attempt to fix startsWith.

Also, actually fail the workflow in the failing case.

BenHenning committed:
Update static_checks.yml

Use correct quotes for release branch string.

BenHenning committed:
Update static_checks.yml

Add additional context to error case to make it easier to understand why the check fails.

seanlip commented:
Ack, thanks. I saw the note!

BenHenning committed:
Update static_checks.yml

Clarify workflow names & messages.

jcqli reviewed (approved):
LGTM

BenHenning commented:
Thanks @jcqli!
